internal/app/machined/pkg/controllers/runtime/fs_scrub.go

```diff
@@ -0,0 +1,244 @@
+		}
+	}
+
+	for item := range volumes {
+		vol := item.TypedSpec()
+
+		if vol.Phase != block.VolumePhaseReady {
+			continue
+		}
+
+		if vol.Filesystem != block.FilesystemTypeXFS {
+			continue
+		}
+
+		volumeConfig, err := safe.ReaderGetByID[*block.VolumeConfig](ctx, r, item.Metadata().ID())
+		if err != nil {
+			return fmt.Errorf("error getting volume config: %w", err)
+		}
+
+		mountpoint := volumeConfig.TypedSpec().Mount.TargetPath
+
+		var period *time.Duration
+
+		for fs := range cfg.All() {
+			if fs.TypedSpec().Mountpoint == mountpoint {
+				period = &fs.TypedSpec().Period
+			}
+		}
+
+		_, ok := ctrl.schedule[mountpoint]
+
+		if period == nil {
+			logger.Warn("!!! scrub !!! not in config, descheduling", zap.String("mountpoint", mountpoint))
+
+			if ok {
+				ctrl.cancelScrub(mountpoint)
+			}
+
+			continue
+		}
+
+		if !ok {
+			firstTimeout := time.Duration(rand.Int64N(int64(period.Seconds()))) * time.Second
+			logger.Warn("!!! scrub !!! firstTimeout", zap.Duration("firstTimeout", firstTimeout))
+
+			// When scheduling the first scrub, we use a random time to avoid all scrubs running in a row.
+			// After the first scrub, we use the period defined in the config.
+			cb := func() {
+				logger.Warn("!!! scrub !!! ding", zap.String("path", mountpoint))
+				ctrl.c <- mountpoint
+				ctrl.schedule[mountpoint].timer.Reset(ctrl.schedule[mountpoint].period)
+			}
+
+			ctrl.schedule[mountpoint] = scrubSchedule{
+				period: *period,
+				timer:  time.AfterFunc(firstTimeout, cb),
+			}
+
+			logger.Warn("!!! scrub !!! scheduled", zap.String("path", mountpoint), zap.Duration("period", *period))
+		} else if ctrl.schedule[mountpoint].period != *period {
+			// reschedule if period has changed
+			logger.Warn("!!! scrub !!! reschedule", zap.String("path", mountpoint), zap.Duration("period", *period))
+
+			ctrl.schedule[mountpoint].timer.Stop()
+			ctrl.schedule[mountpoint].timer.Reset(*period)
+			ctrl.schedule[mountpoint] = scrubSchedule{
+				period: *period,
+				timer:  ctrl.schedule[mountpoint].timer,
+			}
+		}
+	}
+
+	return err
+}
+
+func (ctrl *FSScrubController) cancelScrub(mountpoint string) {
+	ctrl.schedule[mountpoint].timer.Stop()
+	delete(ctrl.schedule, mountpoint)
+}
+
+func (ctrl *FSScrubController) runScrub(mountpoint string, opts []string) error {
+	args := []string{"/usr/sbin/xfs_scrub", "-T", "-v"}
+	args = append(args, opts...)
+	args = append(args, mountpoint)
+
+	r := process.NewRunner(
+		false,
+		&runner.Args{
+			ID:          "fs_scrub",
+			ProcessArgs: args,
+		},
+		runner.WithLoggingManager(ctrl.Runtime.Logging()),
+		runner.WithEnv(environment.Get(ctrl.Runtime.Config())),
+		runner.WithOOMScoreAdj(-999),
+		runner.WithDroppedCapabilities(constants.XFSScrubDroppedCapabilities),
+		runner.WithPriority(19),
+		runner.WithIOPriority(runner.IoprioClassIdle, 7),
+		runner.WithSchedulingPolicy(runner.SchedulingPolicyIdle),
+	)
+
+	return r.Run(func(s events.ServiceState, msg string, args ...any) {})
```

Comment:
I wonder if we should have a way to cancel running scrub process?

Reply:
We can add a channel to signal this once document is deleted. However I haven't studied whether or not it's okay to abort the process and does it terminate safely